### DISCUSSION TITLE
Débat Arcade

### DISCUSSION THREAD

**Discussion by Pebloop** [TempoDev/Arcade-Toulouse#9] [Ideas]
Bonjour, il faut travailler sur quoi faire et comment améliorer les interfaces de l'arcade.
Il faut définir ce que contient scène et comment intéragir entre le jeu, le core et la librairie graphique.

Merci de participer et de donner vos avis ou de préciser ce que vous ne comprenez pas, sinon on fera l'archi sans vous tant pis pour vous.

**Premierement, la lib game interagie avec une scene, il serait bon de définir ce que l'on peut trouver dans cette classe.**

**Comment by fenikssc** [comment]
Dans la scène je viendrais mettre un ensemble d'entités, donc ajouter les méthodes addEntity, getEntity
Ça viendrait à créer une autre classe Entity qui contiendrait ce dont une entité a besoin hp, position etc..

**Comment by Pebloop** [comment]
Oui, je propose d'ajouter un attribut `m_entities` dans la scène :

```cpp
std::unordered_map<std::string, arcade::entity> m_entities;
```

Ainsi que les méthodes : 
```cpp
arcade::Entity &getEntity(std::string name);
std::unordered_map<std::string, arcade::entity> &getEntities();
void addEntity(arcade::Entity &entity);
void removeEntity(std::string name);
```

et donc une classe `class Entity`.

**Reply by smbss1** [reply]
Je ne suis pas d'accord sur le fait que Scene soit une classe au lieu d'une interface. Je pense que c'est mieux d'avoir que des interfaces en commun pour ne pas avoir de problème avec les noms de variables plus tard.

**Reply by Pebloop** [reply]
Issue créée : https://github.com/TempoDev/Arcade-Toulouse/issues/11

**Comment by nasso** [comment]
Je pense qu'`Entity` devrait plutôt être une interface afin de donner le plus de libertés à l'implémentation:

```cpp
class IEntity {
public:
    virtual ~IEntity() = default;
    
    virtual void addComponent(const std::string& name, void* data) = 0;
    virtual bool hasComponent(const std::string& name) const = 0;
    virtual void* getComponent(const std::string& name) = 0;
};
```

**Reply by fenikssc** [reply]
Je ne comprends pas l'idée de "addComponent" alors qu'on aurait déjà un addEntity dans la scene

**Reply by Pebloop** [reply]
Bah les composants sont ce que contient une entitée, mais les entitées sont juste des objets de ton jeu, genre une barre de vie, le joueur, etc qui sont contenu dans la scène :

En gros ta scène contient des entitées (des objets : personnage, barre de vie, ...) qui contiennent eux-même des composants (un sprite, des points de vie, des bruitages, ...).

**Comment by Pebloop** [comment]
Je ne vois pas de problème mais il faut aussi définir de façon commune ce qu'est un sprite, un son, les interactions claviers, etc, puisqu'ils faut que tout le monde puisse les lire et les interpréter dans game.
Donc avant de définir entity, je pense qu'il faut se pencher sur ça.

**Reply by Pebloop** [reply]
Discussion sur le sujet ouverte : #10
